text/2017-10-06-collector-api.md

```diff
@@ -0,0 +1,147 @@
+# Summary
+
+This RFC introduces the `Collector` API for supporting diversified use cases.
+
+
+# Motivation
+
+As the Crossbeam project is growing, users want more diversified API than the existing one.
+
+First, majority of users will be happy with the current API, which interacts with the default
+singleton garbage collector.
+
+Second, some users may want to create their own garbage collector, and use the collector via a
+*handle* to it. Since the current API provides only one garbage collector, Crossbeam currently does
+not support this use case.
+
+Lastly, some users want to embed a garbage collector in another systems library, e.g. memory
+allocator and thread management environment.  Note that it is often a part of the standard library,
+thus this use case requires (at least a part of) Crossbeam to be implemented in the `#[no_std]`
+environment. Likewise, Crossbeam currently does not support this use case.
+
+In order to support these diversified use cases, this author proposes to introduce the **`Collector`
+API**.
+
+
+
+# Detailed design
+
+FYI, this RFC is fully implemented in [this
+branch](https://github.com/jeehoonkang/crossbeam-epoch/tree/handle). You can find a discussion on
+this branch in [this PR](https://github.com/crossbeam-rs/crossbeam-epoch/pull/21).
+
+
+
+## `Collector` API
+
+In this layer of API, the public struct `Collector` is a general-purpose garbage collector with its
+own global epoch, and the public struct `Handle` is an abstraction of a participant in a collector.
+`Collector` is a counted reference to a the global data for garbage collection (`Global`), and
+`Handle` is a tuple of a counted reference to a global data and a local data (`Local`):
+
+```rust
+pub struct Collector(Arc<Global>);
+
+pub struct Handle {
+    global: Arc<Global>,
+    local: Local,
+}
+```
+
+You can create a new collector by creating a new `Global` data, and add a handle to a collector by
+sharing its `Global` data.  Using a `Handle`, you can create a `Scope` and pass it to a given
+function:
+
+```rust
+impl Collector {
+    pub fn new() -> Self;
+    pub fn handle(&self) -> Handle;
+}
+
+impl Handle {
+    pub fn pin<F, R>(&self, f: F) 
+    where F: FnOnce(&Scope) -> R {
+        self.local.pin(&self.global, f)
+    }
+
+    ... // other methods
+}
+```
+
+Note that `Handle::pin()`, on contrary to `Local::pin()`, does not require a reference to a
+`Global`, because `Handle` already has a reference.
+
+See [the
+implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/collector.rs) for
+more details.
+
+
+## The default garbage collector API
+
+There will be the default collector and per-thread handle to the default collector:
+
+```rust
+lazy_static! { pub GLOBAL: Global = Global::new(); }
+thread_local! { pub HANDLE: Handle = Handle::new(); }
```

Comment:
Do we still need those `Global` and `Handle` types? Why don't we have this:

```rust
lazy_static! { pub COLLECTOR: Collector = Collector::new(); }
thread_local! { pub HANDLE: Handle = COLLECTOR.handle(); }
```

Reply:
The `DefaultHandle` has one less pointer indirection, since it's contained in the lazy_static. With `Arc` you have to get to the static first, and then to `ArcInner`. We could replace them if there is no demonstrable overhead to doing this.

Reply:
This pointer indirection (the `Arc<Collector>`) probably won't cost much because we have to fetch the contents of the `Handle` into cache anyway.

Note that `lazy_static!` contains a `Once` and performs an atomic `SeqCst` load on every access (`deref`). It also [boxes](https://github.com/rust-lang-nursery/lazy-static.rs/blob/master/src/lazy.rs#L23) the value (unless you use nightly) so there's still a pointer indirection. So `lazy_static!` may in fact be even worse.

Reply:
Hm, I did forget that `Handle` contains an `Arc<Global>` member, so we don't have to reach through the static. In that case I support the idea to completely remove `DefaultHandle` and replcae it with the new types.

Reply:
I updated the RFC and the corresponding PR.  Please have a look!